README.md
# rpg-python

# Run main_rpg.py to start game. 

# Welcome to a simple text based RPG where a hero fends off three enemies.

Below are User Story requirements of the project: 

(done) Commits
(5 points): As a data analyst, I want to make at least five commits on GitHub with descriptive messages.

(done) Story
(2.5 points): As a user, I want an engaging story to be told using print() statements.

(done) Hero/Enemy Vars 
(2.5 points): As a data analyst, I want to create a Hero variable and set it equal to a Dictionary.
(2.5 points): As a data analyst, I want to create THREE Enemy variables and set each equal to a Dictionary.

(done) Attacks
(5 points): As a user, I want the ability to select my Hero’s attacks based on my input.
(5 points): As a user, I want the Enemy’s attack to be chosen at random.

(done) Health/Damage
(5 points): As a user, I want my Hero or Enemy’s health to decrease based on the power of the successful attack.
(2.5 points): As a user, I want the results of each attack to be printed to the terminal, so I know how much health both opponents have remaining.
(5 points): As a user, I want the Hero and Enemy to continue attacking each other until one's health reaches zero.

(done) Loot
(10 points): As a user, I want to be able to “loot” defeated enemies, which will include:
    · Adding the Enemy’s equipment to the hero character’s equipment Set
    · Adding the Enemy’s gold, silver, or copper coins to the appropriate value in the Hero’s coins Dictionary

(done) Ending
(5 points): As a data analyst, I want the game to conclude when the Hero’s health reaches 0, or when all three Enemies have been defeated.

(done) Setup
(2.5 points): As a data analyst, I want my run_game() function to call my other functions in a logical order that will determine game flow.
(5 points): As a data analyst, I want all of my functions to have a Single Responsibility. Remember, each function should do just one thing!

Bonus
(done)(7.5 points): As a user, I want to increase my Hero’s level after defeating an Enemy, which will include:
    · Incrementing my Hero’s level by 1
    · Adding a new attack of my choosing to my Hero's Attack Tuple.
(done)(5 points): As a user, I want a challenges phrase to be randomly selected and printed when my Hero attacks, and a plea phrase to be randomly selected and printed when my Hero is attacked.




<!-- 5 commits
engaging story

hero dict
enemy dict

select hero attack based on input
random enemy attack

hero and enemies battle until one dies
    print damage outputs
    at death of enemy:
        loot equipment
        loot monies
        
        bonus:
            increase level by 1
            add new attack name + damage var
            print hero challenge on attack and plea on defense

ending
    if hero health is 0 
    or if all enemies defeated -->




change_log.txt
#### wishlist
# add color and speed to text outputs
# modify enemy to say their phrases on attack and defense

#### done - commited
# summary:
# refactored: hero_attack_choice, battle, random_enemy_attack functions

# fixed loot_equip_and_monies function

# added functions: new_hero_attack, hero_level_up, outro, hero_phrase_on_attack, hero_phrase_on_defend


#### done - commited
# added following functions: intro, start_or_not, battle, loot_equip_and_monies, hero_attack_choice. added some story line aspects and print outs to terminal

tests.py
import hero_character
import enemy_characters
import functions
import random
import sys, time


# print(f"Press 1 to use {hero_character.hero['attacks'][0][0]} which deals {hero_character.hero['attacks'][0][1]} damage.")

# print(enemy_characters.enemy_list[0])

# print(f"{enemy_characters.enemy1['attacks']}")
# print(random.choice(enemy_characters.enemy1['attacks'])[1])

# a_random_attack = random.choice(enemy_characters.enemy3['attacks1'])[1]
# print(a_random_attack)
# attack_damage = random.randint(a_random_attack[0], a_random_attack[1])
# print(attack_damage)

# being = {
#     'name': 'xyz',
#     'coins': {
#         'copper': random.randint(0, 100),
#         'silver': random.randint(0, 100)
#     },
#     'attacks': (                                            # TUPLE of TUPLES (each tuple will have attack_name: String, attack_power: Integer)
#         ('Barbarous Slice', (11, 19)),
#         ('Seeking Blade', (29, 39)),
#         ('Dragon Slash', (44, 56))
#     )
# }

# being2 = {
#     'name': 'xyz',
#     'coins': {
#         'copper': random.randint(0, 100),
#         'silver': random.randint(0, 100)
#     }
# }

# print(f"Being has Copper: {being['coins']['copper']} | Silver: {being['coins']['silver']}.")
# print(f"Being 2 has {being2['coins']['copper']} copper.")
# being['coins']['copper'] = being['coins']['copper'] + (being2['coins']['copper'])
# if being['coins']['copper'] > 100:
#             being['coins']['copper'] = being['coins']['copper'] - 100
#             being['coins']['silver'] = being['coins']['silver'] + 1
# being2['coins']['copper'] = 0
# print(f"Being has {being['coins']['copper']} copper.")
# print(f"Being has {being['coins']['silver']} silver.")
# print(f"Being 2 has {being2['coins']['copper']} copper")

# print(random.choice(enemy_characters.enemy1['phrases']['on_attack']))

# print(f"Equipment: {hero_character.hero['equipment']}")

# functions.collected_equip(enemy_characters.enemy1)


# def new_attack():
#     new_attack_name = input('Enter new attack name: ')
#     new_attack_damage = random.randint(15, 75)

#     temp_attack_list = []
#     temp_attack_list.append(new_attack_name)
#     temp_attack_list.append(new_attack_damage)
#     temp_attack_list = tuple(temp_attack_list)
#     attack_list = list(being['attacks'])
#     attack_list.append(temp_attack_list)
#     being['attacks'] = tuple(attack_list)
#     return being['attacks']

# new_attack()

# def random_attack_number():
#     random_attack = random.choice(being['attacks'])
#     enemy_attack_name = random_attack[0]
#     print(random_attack)
#     print(enemy_attack_name)
#     enemy_attack = random.randint(random_attack[1][0], random_attack[1][1])
#     print(enemy_attack)

# random_attack_number()



# def slow_print(string_to_slow_print, delay_time):
#     for letter in string_to_slow_print:
#         sys.stdout.write(letter) # queues up the letter to be printed
#         sys.stdout.flush() # 'flushes' buffer and prints to terminal
#         time.sleep(delay_time) # pauses program for a given amount of time
#         print('')

# slow_print("\033[1;33;48m" + "Hello world, but slowly", 0.05)
# slow_print("Hello world, but slowly", 0.04)

# colored text prints
# utilizing ANSI escape codes
# \033[1;33;48m = Bold Yellow on Black Background
# \033[1;31;48m = Bold Red on Black Background
# \033[1;32;48m = Bold Green on Black Background

functions.py

import random
import hero_character
import enemy_characters 

# functions to be used in main run_game() function

# intro story 
def intro():
    print("\n#############################\n  ***WHOOOSSHHH... POP!*** \n#############################\n\n ")
    print("You hear what appears to be a mage utter under his breath: \n\
'Let's hope our hero can do it this time...' \n\
\nMage: \n\
'*Uhummm.*' \n'Hello young champion! You are here in a training simulation in the Elemental Plains where you will gain worthwhile experience, preparing you for the defense of our world.' \n\
'You must slay these elementals before us, or we will perish. Sorry to tell you, but I can't help you as I'm all out of magic.'")
    player_name = input("\nRemind me, what might your name be?: ")
    hero_character.hero['name'] = player_name

# player deciding whether to start the game or not
def start_or_not():
    global advt_start
    advt_start = False
    while advt_start == False:
        start_or_not = input(f"\nWell {hero_character.hero['name']}, shall we begin? Enter yes/no: ")
        if start_or_not == "yes":
            print("\nLovely! Do me proud and best of luck!")
            print(f"\nYou are starting with {hero_character.hero['health']} health points.")
            advt_start = True
            return advt_start
        elif start_or_not == "no":
            print("\nGuess we'll just wait here until we get beaten.\nYou and the Mage have succumb to the enemies before you.\n\nGame Over.\n")
            advt_start = False
            return advt_start


# during attack, hero or enemy health will decrease based on power of attack
# must show these results to terminal
# hero and enemy must continue to attack until one reaches 0 health
def battle(hero, enemy):
    while advt_start == True:
        while hero['health'] > 0 and enemy['health'] > 0: 
            enemy_intro(enemy)
            # hero attacks first
            hero_attack = hero_attack_choice()
            enemy['health'] -= hero_attack[1]
            if enemy['health'] <= 0:
                print(f"\n'{hero_phrase_on_attack(hero)}'You deal {hero_attack[1]} damage to {enemy['name']} with {hero_attack[0]} and it succumbs to your might!")
                print(f"\nYou have defeated {enemy['name']}!")
                loot_equip_and_monies(hero, enemy)
                hero_level_up(hero)
                new_hero_attack(hero)
                return
            else:
                print('\n~~~Battle~~~')
                print(f"'{hero_phrase_on_attack(hero)}' You deal {hero_attack[1]} damage to {enemy['name']} with {hero_attack[0]}! {enemy['name']} has {enemy['health']} health remaining.")

            # enemy attacks next if health > 0 
            if enemy['health'] > 0:
                enemy_attack = random_enemy_attack(enemy)
                hero['health'] -= enemy_attack[1]
                
                # is hero alive after attack? 
                if hero['health'] <= 0:
                    print(f"{enemy['name']} deals {enemy_attack[1]} damage to you with {enemy_attack[0]}! '{hero_phrase_on_defend(hero)}' You have 0 health remaining.")
                else:
                    print(f"{enemy['name']} deals {enemy_attack[1]} damage to you with {enemy_attack[0]}! '{hero_phrase_on_defend(hero)}' You have {hero['health']} health remaining. ")
            

            if enemy['health'] > 0 and hero['health'] <= 0:
                print(f"\nYou have been defeated by {enemy['name']}. You have failed.")
                return
        return
    else:
        return


# function introducing enemy
def enemy_intro(enemy):
    if enemy['health'] == enemy['max_health']:
        print(f"\nMage: \n\
Before you stands {enemy['name']} who has {enemy['health']} health. Get ready!")
        print(f"You have {hero_character.hero['health']} health points remaining.")
    return

# as a user, select hero attack based on input
def hero_attack_choice():
    while advt_start == True:
        print(f"\nYou have the following abilities:")
        counter = 1
        for attack in hero_character.hero['attacks']:
            print(f"Enter {counter} to use {attack[0]} which deals {attack[1]} damage.")
            counter += 1

        '''hero_attack_choice = int(input(f"Which ability do you use: ") or 0)
        
        while hero_attack_choice < 1 or hero_attack_choice > len(hero_character.hero['attacks']):
            hero_attack_choice = int(input(f"Enter {list(range(1, len(hero_character.hero['attacks']) + 1))} to select an attack! Ability choice: ".replace("[", "").replace("]", "")) or 0)

        hero_attack = hero_character.hero['attacks'][hero_attack_choice - 1][1]
        return hero_attack'''

        # refactoring to allow for user input error handling
        # don't convert to integer
        hero_attack_choice = input(f"Which ability do you use: ")
        
        # check if isdigit FIRST
        while not hero_attack_choice.isdigit() or int(hero_attack_choice) < 1 or int(hero_attack_choice) > len(hero_character.hero['attacks']):
            hero_attack_choice = input(f"Enter {list(range(1, len(hero_character.hero['attacks']) + 1))} to select an attack! Ability choice: ".replace("[", "").replace("]", ""))
        
        # now convert to integer
        hero_attack = hero_character.hero['attacks'][int(hero_attack_choice) - 1][1]
        # pull out hero attack name
        hero_attack_name = hero_character.hero['attacks'][int(hero_attack_choice) -1][0]
        return hero_attack_name, hero_attack
        


# for bonus: random hero phrase on_attack
def hero_phrase_on_attack(hero):
    on_attack = random.choice(hero['phrases']['on_attack'])
    return on_attack


# for bonus: random hero phrase on_attack
def hero_phrase_on_defend(hero):
    on_defend = random.choice(hero['phrases']['on_defend'])
    return on_defend

# have enemy attack be selected at random
    # may add at a later time: random enemy phrase on attack and defend
def random_enemy_attack(enemy):
    random_attack = random.choice(enemy['attacks'])
    enemy_attack_name = random_attack[0]
    enemy_attack = random.randint(random_attack[1][0], random_attack[1][1])
    return enemy_attack_name, enemy_attack


# hero will loot enemies including equipment and coins/monies
def loot_equip_and_monies(hero, enemy):
    if hero['health'] > 0 and enemy['health'] <= 0:
        # equipment loot
        hero['equipment'].update(enemy['equipment'])
        # equipment_collected = collected_equip(enemy)
        print(f"\nEquipment collected: {collected_equip(enemy)}")
        
        # monies - copper, max 100
        hero['coins']['copper'] = hero['coins']['copper'] + enemy['coins']['copper']
        if hero['coins']['copper'] > 100:
            hero['coins']['copper'] = hero['coins']['copper'] - 100
            hero['coins']['silver'] = hero['coins']['silver'] + 1
        # silver, max 100
        hero['coins']['silver'] = hero['coins']['silver'] + enemy['coins']['silver']
        if hero['coins']['silver'] > 100:
            hero['coins']['silver'] = hero['coins']['silver'] - 100
            hero['coins']['gold'] = hero['coins']['gold'] + 1
        # gold
        hero['coins']['gold'] = hero['coins']['gold'] + enemy['coins']['gold']

        print(f"Coins collected: Gold: {enemy['coins']['gold']}, Silver: {enemy['coins']['silver']}, Copper: {enemy['coins']['copper']}")
        print(f"Coin pouch: {hero['coins']['gold']}g {hero['coins']['silver']}s {hero['coins']['copper']}c")
        
        enemy['coins']['copper'] = 0
        enemy['coins']['silver'] = 0
        enemy['coins']['gold'] = 0
    else:
        hero['equipment'].clear()
        hero['coins']['copper'] = 0
        hero['coins']['silver'] = 0
        hero['coins']['gold'] = 0


# collected equipment output text
def collected_equip(enemy):
    for e in enemy['equipment']:
        return e


# for bonus: hero levels up on enemy defeat
def hero_level_up(hero):
    hero['level'] += 1
    print(f"\nYou have gained a level! You are now Level {hero['level']}")


# for bonus: new hero attack input
def new_hero_attack(hero):
    new_hero_attack_name = input("You've gained a new skill. Please enter a new attack name: ")
    new_hero_attack_damage = random.randint(15, 75)
    print(f'Your new attack {new_hero_attack_name} deals a randomly generated {new_hero_attack_damage} damage.')

    temp_attack_list = []
    temp_attack_list.append(new_hero_attack_name)
    temp_attack_list.append(new_hero_attack_damage)
    temp_attack_list = tuple(temp_attack_list)

    attack_list = list(hero['attacks'])
    attack_list.append(temp_attack_list)
    hero['attacks'] = tuple(attack_list)
    return hero['attacks']


# create ending story if hero alive or dead
def outro(hero):
    while advt_start == True:
        if hero['health'] <= 0:
            print(f"\nMage: \n\
Eh, don't worry about it {hero['name']}! You're not the first to not complete this training exercise.\
\nWe'll come back to these elementals when you are better prepared!\
\nI'll teleport you back now, get some rest.")
            print('\n#############################\n  ***WHOOOSSHHH... POP!*** \n#############################')
            print('\nGAME OVER.\n')
            return
        if hero['health'] > 0:
            print(f"\nMage: \n\
Well done {hero['name']}! You have completed your Elemental Trials with {hero['health']} health remaining! \nMy energy is restored and I will now teleport you back to the training grounds for your next session. See you there! ;)")
            print('\n#############################\n  ***WHOOOSSHHH... POP!*** \n#############################')
            print('\nCongratulations, you have completed the game.')
            print('\nGame Over.\n')
            return
    return

hero_character.py

# main hero character

hero = {                                                    # DICT{} all encompassing
    'name': 'test',                                             # String
    'level': 1,                                             # Integer
    'health': 350,                                          # Integer
    'equipment': {"Father's Broadsword"},                   # SET{} of strings
    'attacks': (                                            # TUPLE of TUPLES (each tuple will have attack_name: String, attack_power: Integer)
        ('Barbarous Slice', 25),
        ('Seeking Blade', 50),
        ('Dragon Slash', 75)
    ),        
    'coins': {                                              # DICT{}
        'copper': 0,                                        # Integer
        'silver': 0, 
        'gold': 0, 
    },
    'phrases': {
        'on_attack': ["and...stay down!", 'SMASH!', 'Feel my RAGEEEE!'],     # LIST[] of Strings
        'on_defend': ["Yikes!", 'Ouch!', 'Zoinks!']   # LIST[] of Strings
    }
}


enemy_characters.py

import random

# 3 enemy characters

enemy1 = {                                                  # DICT{} all encompassing
    'name': 'Marble, the Earthbreaker',                       # String
    'level': 1,                                             # Integer
    'health': 160,
    'max_health': 160,                                          # Integer
    'equipment': {"Savage Greatboots of the Mountain"},     # SET{} of strings
    'attacks': (                                            # TUPLE of TUPLES (each tuple will have attack_name: String, attack_power: Integer)
        ('Rock Throw', (11, 19)),
        ('Stone Spike', (29, 39)),
        ('Earthquake', (44, 56))
    ),  
    'coins': {                                              # DICT{}
        'copper': random.randint(0, 100),                   # Integer
        'silver': random.randint(0, 100), 
        'gold': random.randint(1, 2), 
    },
    'phrases': {
        'on_attack': ['The Earth crumbles under you!', "You're stuck between a rock and a hard place!", 'Mama said knock you out!'],     # LIST[] of Strings
        'on_defend': ['That will leave a mark...', 'Noooooo', 'I was not expecting that...']   # LIST[] of Strings
    }
}

enemy2 = {                                                  # DICT{} all encompassing
    'name': 'Hurricane Voxis',                              # String
    'level': 1,                                             # Integer
    'health': 160,
    'max_health': 160,                                          # Integer
    'equipment': {"Stormfury Chainmail Chestplate"},        # SET{} of strings
    'attacks': (                                            # TUPLE of TUPLES (each tuple will have attack_name: String, attack_power: Integer)
        ('Air Blast', (11, 19)),
        ('Wind Cut', (29, 39)),
        ('Hurricane', (44, 56))
    ),            
    'coins': {                                              # DICT{}
        'copper': random.randint(0, 100),                   # Integer
        'silver': random.randint(0, 100), 
        'gold': random.randint(1, 2), 
    },
    'phrases': {
        'on_attack': ['Almighty wind!', 'Huzzah!', 'You will get blown away!'],     # LIST[] of Strings
        'on_defend': ['Ruh roh...', 'Ease up on me will ya?', 'I will get you for that one!']   # LIST[] of Strings
    }
}

enemy3 = {                                                  # DICT{} all encompassing
    'name': 'Galvatron, Lord of Lightning',                      # String
    'level': 1,                                             # Integer
    'health': 300,
    'max_health': 300,                                          # Integer
    'equipment': {'Infinite Light Necklace'},               # SET{} of strings
    'attacks': (                                            # TUPLE of TUPLES (each tuple will have attack_name: String, attack_power: Integer)
        ('Electric Zap', (11, 19)),
        ('Electric Shock', (29, 39)),
        ('Lightning Storm', (44, 56))
    ),
    'coins': {                                              # DICT{}
        'copper': random.randint(0, 100),                   # Integer
        'silver': random.randint(0, 100), 
        'gold': random.randint(3, 7), 
    },
    'phrases': {
        'on_attack': ['Welcome to the THUNDERDOME', 'Zip Zap!', 'Electrification!'],     # LIST[] of Strings
        'on_defend': ['Ooooof!', 'How did you do that?!', 'Oh jeez, oh no!']   # LIST[] of Strings
    }
}
